renderer_cocoa/lib/Cocoa_stubs.c

```diff
@@ -0,0 +1,780 @@
+
+  View *view = View_val(view_v);
+
+  CGFloat x = Double_val(x_v);
+  CGFloat y = Double_val(y_v);
+  CGFloat w = Double_val(w_v);
+  CGFloat h = Double_val(h_v);
+
+  NSRect rect = NSMakeRect(x, y, w, h);
+  [view setFrame:rect];
+
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value ml_NSView_setBorderWidth(value view_v, value width_v)
+{
+  CAMLparam2(view_v, width_v);
+
+  View *view = View_val(view_v);
+
+  [view.layer setBorderWidth:Int_val(width_v)];
+
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value ml_NSView_setBorderColor(value view_v, value red_v, value green_v, value blue_v, value alpha_v)
+{
+  CAMLparam5(view_v, red_v, blue_v, green_v, alpha_v);
+
+  View *view = View_val(view_v);
+
+  CGFloat red = Double_val(red_v) / 255;
+  CGFloat blue = Double_val(blue_v) / 255;
+  CGFloat green = Double_val(green_v) / 255;
+  CGFloat alpha = Double_val(alpha_v);
+
+  [view setWantsLayer:YES];
+  [view.layer setBorderColor:[[NSColor colorWithRed:red green:green blue:blue alpha:alpha] CGColor]];
+
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value ml_NSView_setBackgroundColor(value view_v, value red_v, value green_v, value blue_v, value alpha_v)
+{
+  CAMLparam5(view_v, red_v, blue_v, green_v, alpha_v);
+
+  View *view = View_val(view_v);
+
+  CGFloat red = Double_val(red_v) / 255;
+  CGFloat blue = Double_val(blue_v) / 255;
+  CGFloat green = Double_val(green_v) / 255;
+  CGFloat alpha = Double_val(alpha_v);
+
+  [view setWantsLayer:YES];
+  [view.layer setBackgroundColor:[[NSColor colorWithRed:red green:green blue:blue alpha:alpha] CGColor]];
+
+  CAMLreturn(Val_unit);
+}
+
+// NSButton
+CAMLprim value ml_NSButton_make()
+{
+  CAMLparam0();
+
+  Button *btn = nil;
+  btn = [Button new];
+  [ml_Views_all addObject:btn];
+
+  CAMLreturn(Val_Button(btn));
+}
+
+CAMLprim value ml_NSButton_memoize(value id_v, value btn_v)
+{
+  CAMLparam2(id_v, btn_v);
+  Button *btn = Button_val(btn_v);
+
+  [ml_Views setObject:btn forKey:@(Int_val(id_v))];
+
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value ml_NSButton_free(value id_v)
+{
+  CAMLparam1(id_v);
+
+  [ml_Views removeObjectForKey:@(Int_val(id_v))];
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value ml_NSButton_setFrame(value btn_v, value x_v, value y_v, value w_v, value h_v)
+{
+  CAMLparam5(btn_v, x_v, y_v, w_v, h_v);
+
+  Button *btn = Button_val(btn_v);
+
+  CGFloat x = Double_val(x_v);
+  CGFloat y = Double_val(y_v);
+  CGFloat w = Double_val(w_v);
+  CGFloat h = Double_val(h_v);
+
+  NSRect rect = NSMakeRect(x, y, w, h);
+  [btn setFrame:rect];
+
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value ml_NSButton_setCallback(value btn_v, value callback_v)
+{
+  CAMLparam2(btn_v, callback_v);
+
+  Button *btn = Button_val(btn_v);
+  value callback = callback_v;
+
+  caml_register_global_root(&callback);
+
+  [btn onClick:^{
+    NSLog(@"sent");
+    caml_callback(callback, Val_unit);
+  }];
+
+  CAMLreturn(Val_Button(btn));
+}
+
+CAMLprim value ml_NSButton_setTitle(value btn_v, value str_v)
```

Comment:
I'm for removing everything that's not bare minimum from this PR, too.

Reply:
Dropped as much as possible, but we need the title for demo purposes. 😁